go/enclave/rollupchain/rollup_chain.go

```diff
@@ -270,24 +270,45 @@ func (rc *RollupChain) findRoundWinner(receivedRollups []*obscurocore.Rollup, pa
 	return headRollup
 }
 
-func (rc *RollupChain) process(rollup *obscurocore.Rollup, stateDB *state.StateDB, depositTxs []*common.L2Tx) (gethcommon.Hash, []*types.Receipt, []*types.Receipt) {
-	txReceipts := evm.ExecuteTransactions(rollup.Transactions, stateDB, rollup.Header, rc.storage, rc.obscuroChainID, 0)
-	if len(rollup.Transactions) != len(txReceipts) {
-		panic("Sanity check. All transactions that are included in a rollup must be executed and produce a receipt.")
+func (rc *RollupChain) process(rollup *obscurocore.Rollup, txs obscurocore.L2Txs, stateDB *state.StateDB) (gethcommon.Hash, obscurocore.L2Txs, []*types.Receipt, []*types.Receipt) {
+	var successfulTransactions obscurocore.L2Txs
+	txReceipts := evm.ExecuteTransactions(txs, stateDB, rollup.Header, rc.storage, rc.obscuroChainID, 0)
+	txReceiptsMap := toReceiptMap(txReceipts)
+	// todo - only transactions that fail because of the nonce should be excluded
+	for _, tx := range txs {
+		_, f := txReceiptsMap[tx.Hash()]
+		if f {
+			successfulTransactions = append(successfulTransactions, tx)
+		} else {
+			log.Info(">   Agg%d: Excluding transaction %d", common.ShortAddress(rc.hostID), common.ShortHash(tx.Hash()))
+		}
+	}
+
+	// always process deposits last, either on top of the rollup produced speculatively or the newly created rollup
+	// process deposits from the fromBlock of the parent to the current block (which is the fromBlock of the new rollup)
+	depositTxs := rc.bridge.ExtractDeposits(
+		rc.storage.Proof(rc.storage.ParentRollup(rollup)),
+		rc.storage.Proof(rollup),
+		rc.storage,
+		stateDB,
+	)
+
+	depositReceipts := evm.ExecuteTransactions(depositTxs, stateDB, rollup.Header, rc.storage, rc.obscuroChainID, len(successfulTransactions))
+	if len(depositReceipts) != len(depositTxs) {
+		log.Panic("Sanity check. Some deposit transactions failed.")
 	}
-	depositReceipts := evm.ExecuteTransactions(depositTxs, stateDB, rollup.Header, rc.storage, rc.obscuroChainID, len(rollup.Transactions))
+
 	rootHash, err := stateDB.Commit(true)
 	if err != nil {
 		log.Panic("could not commit to state DB. Cause: %s", err)
 	}
-	return rootHash, txReceipts, depositReceipts
+	return rootHash, successfulTransactions, txReceipts, depositReceipts
 }
 
 func (rc *RollupChain) validateRollup(rollup *obscurocore.Rollup, rootHash gethcommon.Hash, txReceipts []*types.Receipt, depositReceipts []*types.Receipt, stateDB *state.StateDB) bool {
 	h := rollup.Header
 	if !bytes.Equal(rootHash.Bytes(), h.Root.Bytes()) {
-		// dump := stateDB.Dump(&state.DumpConfig{})
-		dump := ""
+		dump := stateDB.Dump(&state.DumpConfig{})
```

Comment:
Does this dump get expensive as the state grows? No need to optimise now but maybe worth a todo to revisit if you think it could be a hotspot.

Reply:
this branch of code is only executed when the calculated state root doesn't match the root declared in the header. So it doesn't really matter how expensive it is, because it's pretty much game over ( for now)